### PR TITLE
Reorganising documentation into Getting Started, Tutorial and How To

### PR DESCRIPTION
**Pull request by chrishalcrow**
New documentation structure, based on discussions in #2763 and #2327.

**The reorganised docs:**
https://spikeinterface--2778.org.readthedocs.build/en/2778/

Idea is to make the documentation simpler to navigate for new (to si and to coding) users. Main toctree is

- Overview
- Getting Started
- Tutorials
- How to...
- In-depth
- API
- as before...

I deliberately just changed the structure and moved files around, and didn't change the files themselves (except some titles, and descriptions on index pages). Once this overall structure is fixed, we can have other debates about what belongs where, and what is missing.

The tutorials are generated by sphinx-gallery, but contain some "plain" rst files. This works, but the two types are linked from index in two different ways. I think this is fine.

The build looks fine locally, but I'm sure there will be some broken things.

Not sure about the name "In-depth". I quite like "How to...". Opinions welcome on everything!
@h-mayorquin @zm711 @JoeZiminski @alejoe91 @samuelgarcia

### CODE REVIEW

**Comment by zm711** [comment]
image

Just going to post the pic so we can see what it looks like along the side in the actual build.

**Comment by chrishalcrow** [comment]
@JoeZiminski suggested putting the Viewers/Visualising_data section in the Tutorials. Thoughts?

**Comment by zm711** [comment]
They aren't super informative (I mean depth-wise), so I would probably think of them more like "how to". Ie how to visualize your data rather than an in-depth visualization tutorial.

As an aside: For me `In depth` is hard to parse. I have no clue what that means. I think `Modules Documentation` would be clearer unless someone has a better name.

**Comment by JoeZiminski** [comment]
Agree @zm711 'Visualising Data' makes sense in 'How To...'. (maybe renamed to 'Visualise Data').

**Comment by h-mayorquin** [comment]
OK, I gave it some thought to this today.
Now that we have green light by both Sam and Alessio we should try to get a version of this working quickly and then improve after. The **lowest hanging fruit is to not have both the "modules documentation" and the "modules examples gallery* at the same level*. This PR does this and therefore it is better than what we have already.

I would even vote to merge as it is and then we can discuss after as the rest. Another option is to make a PR where that is the only change. What do you think?

That said, my opinion as I have read the documentation here and checked how some standard packages (numpy, scipy, scikit-learn, etc) handle their docs:

* Change "how to get started" section that is nessted in "Getting Started" to "quick start". This follow numpy convention https://numpy.org/doc/stable/user/quickstart.html and also avoids this akward (imo) redundancy. 
* Visualizaing data is too short right now to be its own section. I would put it a section of  "getting started" as a pointer and hopefully later we can expand on it and see where it fits better.
* Again, following Numpy my suggestion for "In-depth" would be "Spikeinterface Fundamentals"
* I don't think that the section of  "Tutorials without a notebook" is a good section.  The neuroxpiel too is a how-to (it solves a specific problem) and I think the "From WaveformExtractor to SortingAnalyzer" can be a section on its own called "updating from legacy" or something like that.
* Lastly, I would really like to not use mearec on the "get started", reading the "get started" I think we can use the generator for that but this can be another PR.

I think that big limitation that we have is the organization the documentation that reflects the package organization (the modules). This is easy for us programmers and a good start but overall I don't think it is a good organization principle. Users care about objects and we should center around those objects and what they make affordable. There are modules that maybe we don't want to document, and we don't need to explain everything about a module.

**Comment by JoeZiminski** [comment]
I agree the distinction between In-Depth and other documentation sections might not be clear. I think 'Module Documentation' is okay as a name but it relies on the reader to know what a 'Module' is.

I've had a look through and quick attempt and catalogue (below). The information in there is super-valuable but contains a mix of different categories of documentation. I think this is in part why it is difficult to name and navigate as a user. Because of this I think it matters less what it's called now and we could have a follow-up plan to reorganise it soon. But happy to keep as 'Module Documentation' or maybe 'API Explanation'  (mirroring [diataxis](https://diataxis.fr/) explanation section) or 'Detailed Reference'.

Based on the below I think this section contains a mix of various documentation categories:
- General API details / explanations. These are things like the idea of the [preprocessing chain concept](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/preprocessing.html#overview-the-chain-concept), [Quality metrics module](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/qualitymetrics.html), [sorting components](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/sortingcomponents.html) philosophy and API. I think all of this could be kept here in a 'API Explanation'  / 'More Details' / 'Advanced Documentation' / 'User Guide' section. Its all extremely useful and doesn't fall into another existing category. 
-  Targeted API details / explanations. These are related to specific public-facing functions, in particular in the [preprocessing](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/preprocessing.html#available-preprocessing) section. These are close to the API docs, just with a little more info, and possibly risk duplicating the API section. For these I might suggest moving them to the actual function docstring, similar to a Numpy/Scipy-style approach where the public-facing docstrings contain a lot of detail / explanation. Then everything is in one place and linked closely to the code.
- There are some things that could be moved / merged with existing how-to guides and tutorials

General overview:

## Core module

I would suggest the below subsections are API reference / developer docs that cover how the API is structured but aren’t linked to any specific functionality or goal. These could be kind of deep API reference / explanation (mostly for developers):

- Recording
- Sorting
- Event
- Snippets
- Recording tools
- Template tools
- LEGACY objects

 These subsections are like 'How To's and could be moved / merged into the 'How To section'
- Handling Probes
- Saving, loading and compression
- Parallel processing and job kwargs
- Manipulating objects:  slicing, aggregating
- Generate Toy objects
- Downloading test datasets

But Sparsity and maybe 'Saving, loading and compression' I'm not sure and are kind of general explanation and useful background.

## Extractors module
 The below sections could go into / be merged with an existing 'Loading data' tutorial or how-to:

- read one recording
- read one sorting
- read one event
- lazy loading

There is an important list of all supported data formats.

## Preprocessing module

Contains a valuable discussion on the chain concept and dtype. This is kind of API explanation.

Then there is an 'Available preprocessing' which has a little more detail on preproecssing functions. Maybe these could be merged into the docstring of an these functions similar to Numpy.

There is a "How to implement “IBL destriping” or “SpikeGLX CatGT” in SpikeInterface[¶](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/preprocessing.html#how-to-implement-ibl-destriping-or-spikeglx-catgt-in-spikeinterface)" that could go to 'How To'.

## Sorters module

It is useful to have a single 'sorting' page that links together everything related to sorting. But I think this could broken up into 3/4 separate articles:
- a general 'Sorting' tutorial introducing the 'wrapper' concept, how to run sorting generally. 
- Details on running sorting in container. either a 'How To' for a kind of explanation article.
- Running several sorters in parallel / spike sorting by group / handling multi-segment recordings /installed sorters: merged into tutorials or standalone how-to
- Supported Spike Sorters (similar to the supported extractors section)
- Contributing guidelines (moved to Contributing)

## Postprocessing module

This could remain as a API explanation article, or be made a tutorial. 

I would suggest the sections that add more detail to the functions are moved into docstring and API docs beefed up here [(available extensions)](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/postprocessing.html#available-postprocessing-extensions)

## Quality Metrics module

This is a very nice detailed deep reference for all the quality metrics. 

## Comparison Module, 

This standalone reference article or tutorial, or maybe the most tutorial-parts taken out for a tutorial and rest left as an information article. 

## Exporters Module

These could be How To:
- Exporting to Phy[¶](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/exporters.html#exporting-to-phy)
- Export a spike sorting report[¶](https://spikeinterface--2778.org.readthedocs.build/en/2778/in_depth/exporters.html#export-a-spike-sorting-report)

## Widgets module

This is kind of a standalone API reference on the available widget backends.

## Curation module

A nice standalone overview of the curation options in SI.

## Sorting Components module

A nice introduction and overview on the Sorting Components concept.

## Motion/direct correction

Again nice standalone article on motion correction options.

**Review by JoeZiminski — APPROVED** [review]
Hey @chrishalcrow this looks great, just a few small suggestions. I agree with @h-mayorquin this could be merged pretty much as-is.

File: doc/get_started/index.rst
Comment (JoeZiminski) [reply doc/get_started/index.rst:11]: actually @h-mayorquin 'quickstart' is good and avoids the redundancy between section name and tutorial name. Maybe 'Quickstart Guide'  or 'Quickside Tutorial' just because 'quickstart' on it's own might be slightly amiguous?

File: doc/get_started/import.rst
Comment (zm711) [reply doc/get_started/import.rst:1]: That discussion will never be solved haha so this doc was the compromise :)

Comment (JoeZiminski) [reply doc/get_started/import.rst:1]: However, this is out of scope for this PR and probably makes sense to revive that thread. I will do so now and link to the above comment.

**Comment by zm711** [comment]
My vote would be these 3 changes and then do a better cleanup either 1) after the conference or 2) hackathon

> Change "how to get started" section that is nessted in "Getting Started" to "quick start". This follow numpy convention https://numpy.org/doc/stable/user/quickstart.html and also avoids this akward (imo) redundancy.

> Visualizaing data is too short right now to be its own section. I would put it a section of "getting started" as a pointer and hopefully later we can expand on it and see where it fits better.

> Again, following Numpy my suggestion for "In-depth" would be "Spikeinterface Fundamentals"

File: doc/get_started/import.rst
Comment (JoeZiminski) [reply doc/get_started/import.rst:1]: Haha yes I had forgotton the extent of the discussion 😄 Okay makes sense

**Comment by JoeZiminski** [comment]
Agree with @h-mayorquin and @zm711 my only thought is 'Visualising Data' could go in 'How To' rather than 'Getting Started' as I think unless you are familiar with SI the article will not make much sense.

File: doc/get_started/install_sorters.rst
Comment (JoeZiminski) [reply doc/get_started/install_sorters.rst:1]: On further consideration I don't think now is the time to do it, and it wouldn't really go with the rest of 'In Depth', so makes sense to ignore this for now.

**Comment by JoeZiminski** [comment]
Sure thanks @zm711, do you think a new issue would make sense?
- #2327 for discussing how we now reorganise how to vs. tutorial
- #2763 can probably be closed soon, maybe spin out more atomic issues on unpinning sphinx version and possibly changing what 'latest' points to pending a little more discussion.
- new issue for reorganisation of 'Module Documentation' ?

**Comment by zm711** [comment]
I think a new issue would be fine. We can have a fresh start!

**Comment by h-mayorquin** [comment]
Thinking twice I don't think we should merge this yet. I think we should remove the generated files out of version control.

**Comment by chrishalcrow** [comment]
Lots of discussion, great!

I'll implement the generally agreed about stuff later, then need to check the version control stuff (good catch @h-mayorquin), and try to clean up references to the old Section names throughout the docs. And we'll talk about what goes where and what can be split-up etc in person in Edinburgh?

Talking of names, seems like "Module documentation" still wins. Let's keep it for now? @alejoe91 suggested "Module by Module" as a name for "Module documentation". I thought it sounded terrible at first, but woke up thinking it was a nice name. 

> I don't think that the section of "Tutorials without a notebook" is a good section. The neuroxpiel too is a how-to (it solves a specific problem) and I think the "From WaveformExtractor to SortingAnalyzer" can be a section on its own called "updating from legacy" or something like that.

Agreed. I wanted to include this as it shows the stylistic constraints if we use sphinx-gallery: we can include basic rst files, but the way they're included means it's not easy to bundle them with the gallery tutorials. So their either in their own section (here in the "Tutorials without a notebook" section) or could be in a subsection, but would appear as a list of links instead of a gallery card. But you're right, for these examples are there are easy fixes: have neuropxiles as a how-to and put "WE to SA" in its own section.

**Comment by JoeZiminski** [comment]
Sounds good @chrishalcrow!

Happy to keep 'Module Documentation' for now. In general my thoughts are to keep the documentation as accessible as possible, as many users may not be very experienced with python. Imagining yourself as a new user with limited python experience navigating the website for the first time, anything with 'Module' in the name is not informative. Obviously the docs will need to get technical in places but not on the front page. So I would advocate for changing it later but happy to keep as is as there are many changes and discussions to come.

In terms of the timeline, for me personally I am away next week but then back on ephys stuff now hopefully for a long a while and was planning to help with the documentation as I realised I've never read it all the way through. So for me it would work to continue discussion and refactorings up until the conference, and this may leave only the bigger things to tackle in person. But happy to proceed in whatever way works best for everyone!

**Review by zm711 — DISMISSED** [review]
Just putting a request changes because I noticed we are clobbering a recent doc PR about one of the quality metrics. See comment. We definitely can't merge until we fix that!

File: doc/modules/qualitymetrics/amplitude_cutoff.rst
Comment (zm711) [inline doc/modules/qualitymetrics/amplitude_cutoff.rst:9]: This whole doc needs to be reverted!

Comment (zm711) [reply doc/modules/qualitymetrics/amplitude_cutoff.rst:9]: I think it was because the PR was opened and merged after this PR. #2756!

**Comment by chrishalcrow** [comment]
Wait! I've done a stupid thing with the analyze neuropixles how to... DON'T MERGE

**Comment by JoeZiminski** [comment]
Thanks Chris! I kind of ignored `.rst` changes for some reason I thought they were autogenerated by sphinx-build but I see from @zm711 the order of merges could affect things. Just checking some changes seen are intended, I wasn't sure if they were intended rewrites or regression overwriting since-merged PRs:
- `core.rst` rewrite
- renaming of `SortingAnalyzer` to `WaveformExtractor` in `exporters.rst`
- section removed from `postprocessing.rst`  
- `qualitrymetrics/amplitude_cutoff.rst` 
- rewrite, some small changes to `plot_4_sorting_analyzer.py`

**Comment by JoeZiminski** [comment]
BTW the title of this PR could be renamed before merge now that the structure is more decided. How about 'Reorganising documentation into Getting Started, Tutorial and How To.'?

**Comment by zm711** [comment]
> Thanks Chris! I kind of ignored `.rst` changes for some reason I thought they were autogenerated by sphinx-build but I see from @zm711 the order of merges could affect things. Just checking some changes seen are intended, I wasn't sure if they were intended rewrites or regression overwriting since-merged PRs:
> 
> * `core.rst` rewrite
> * renaming of `SortingAnalyzer` to `WaveformExtractor` in `exporters.rst`
> * section removed from `postprocessing.rst`
> * `qualitrymetrics/amplitude_cutoff.rst`
> * rewrite, some small changes to `plot_4_sorting_analyzer.py`

I think these could all be regressions! We will need to double check them carefully!

File: doc/get_started/quickstart_files/quickstart_16_0.png
Comment (h-mayorquin) [inline doc/get_started/quickstart_files/quickstart_16_0.png:1]: This png are generated files. I think we should remove them, right?

**Comment by h-mayorquin** [comment]
Let's not forget to squash the commit history for this one!

I checked generated docs and they look OK to me. I don't like "Modules Documentation" because I don't want to structure the documentation around modules. I want to add in depth sections about the `sorting_vector` and the `parallel_processing` for example and the current structure means that everything will be super full in core and kind of empty on extractors for example. But we can discuss this in another PR or issue once I get to formulate this better.

**Comment by zm711** [comment]
@h-mayorquin, Joe just opened a new issue to discuss a deeper rearrange  once this toctree is done. You should read his takes and then add your own there! #2800.

File: doc/get_started/quickstart_files/quickstart_16_0.png
Comment (h-mayorquin) [reply doc/get_started/quickstart_files/quickstart_16_0.png:1]: But they are generated on the read the docs so they will be there for the rst to display. They just don't need to be in version control. Am I wrong?

Comment (chrishalcrow) [reply doc/get_started/quickstart_files/quickstart_16_0.png:1]: These are generated in a different way than the sphinx-gallery files. They're run locally so that we can do big computations and use larger data files. See here: https://github.com/SpikeInterface/spikeinterface/tree/main/examples/how_to
To be honest, they are a pain, but that's a discussion for another PR!

Comment (h-mayorquin) [reply doc/get_started/quickstart_files/quickstart_16_0.png:1]: OK, thanks for the pointer. I forgot that there is yet-another-system of generating files in the docs. Maybe the simplest if to just not change the images every time we run the code (they should be the same!). We should discuss this in another PR or issue though as  this is creating the dreaded super large repo of @samuelgarcia .

File: examples/tutorials/README.rst
Comment (zm711) [inline examples/tutorials/README.rst:6]: New to what? we probably want to say to the library/spikeinterface no? It could be 

new to python
new to coding
new to spikeinterface

File: doc/how_to/index.rst
Comment (zm711) [inline doc/how_to/index.rst:4]:
```suggestion
Guides on how to solve specific, short problems in SpikeInterface. Learn how to...
```
or it could be spikeinterface. Just in case you haven't seen the difference between the two. The overall project is stylized SpikeInterface whereas the package specifically is stylized as spikeinterface.

File: doc/how_to/load_matlab_data.rst
Comment (zm711) [inline doc/how_to/load_matlab_data.rst:1]:
```suggestion
Export MATLAB Data to Binary & Load in SpikeInterface
```
since we removed one gerund might as well switch the other.

**Comment by zm711** [comment]
@chrishalcrow,

Just two typos and then this looks good to me (one comment that doesn't necessarily need to be addressed here).

**Comment by zm711** [comment]
This is good for me. Anything else you wanted @h-mayorquin ?

**Review by h-mayorquin — APPROVED** [review]
No, this is fine for me.

**Review by zm711 — APPROVED** [review]

**Comment by h-mayorquin** [comment]
Thanks for the push for this @chrishalcrow

**Comment by chrishalcrow** [comment]
Thanks for getting this merged - I've been on holiday (I maybe should have mentioned that at some point!!)

Looking forward to much more documentation discussion in the next couple of months!

**Pull request merged**
